<a href="https://colab.research.google.com/github/khadijaMajeed16/flyrank-machine-learning/blob/main/work/notebooks/w03_data_contract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [30]:
!pip install -q duckdb huggingface_hub pyarrow fsspec
from google.colab import userdata
from huggingface_hub import login

HF_TOKEN = userdata.get("HF_TOKEN")

login(token=HF_TOKEN)

print("Logged into Hugging Face successfully!")

Note: Environment variable`HF_TOKEN` is set and is the current active token independently from the token you've just configured.


Logged into Hugging Face successfully!


In [31]:
import duckdb

con = duckdb.connect()

print("DuckDB Connected")

DuckDB Connected


In [32]:
import os

print(os.path.exists("/content/march_2026.parquet"))
print(os.path.getsize("/content/march_2026.parquet"))

True
124233710


In [33]:
import requests

url = "https://huggingface.co/datasets/FlyRank/internship-warehouse/resolve/main/fact_content_daily_performance/month=2026-03/data_0.parquet"

headers = {
    "Authorization": f"Bearer {HF_TOKEN}"
}

r = requests.get(url, headers=headers)

print("Status:", r.status_code)
print("Size:", len(r.content))

Status: 200
Size: 124233710


In [34]:
import duckdb
import fsspec
import huggingface_hub

print("DuckDB:", duckdb.__version__)
print("fsspec:", fsspec.__version__)
print("huggingface_hub:", huggingface_hub.__version__)

DuckDB: 1.3.2
fsspec: 2025.3.0
huggingface_hub: 1.27.0


In [35]:
import os

os.environ["HF_TOKEN"] = HF_TOKEN
os.environ["HF_HUB_TOKEN"] = HF_TOKEN

print("Hugging Face token set for DuckDB/fsspec")

Hugging Face token set for DuckDB/fsspec


In [36]:
import requests

url = "https://huggingface.co/datasets/FlyRank/internship-warehouse/resolve/main/fact_content_daily_performance/month=2026-03/data_0.parquet"

headers = {
    "Authorization": f"Bearer {HF_TOKEN}"
}

r = requests.get(url, headers=headers)

print("Status:", r.status_code)
print("Size:", len(r.content))

with open("/content/march_2026.parquet", "wb") as f:
    f.write(r.content)

print("Downloaded successfully!")

Status: 200
Size: 124233710
Downloaded successfully!


In [37]:
con.sql("""
SELECT *
FROM read_parquet('/content/march_2026.parquet')
LIMIT 5
""").df()

,report_date,client_hash_id,content_hash_id,client_has_gsc,client_has_ga4,gsc_data_available,ga4_data_available,gsc_impressions,gsc_clicks,gsc_sum_position,...,sessions_paid,sessions_ai,ai_chatgpt,ai_perplexity,ai_gemini,ai_copilot,ai_claude,ai_meta,ai_other,scroll_events
0,2026-03-01,client_73cda7b4e4f265ea,content_b7e512995f79d5a6,True,False,True,<NA>,20,0,67,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>
1,2026-03-01,client_73cda7b4e4f265ea,content_05597932fe4da067,True,False,True,<NA>,1,0,0,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>
2,2026-03-01,client_73cda7b4e4f265ea,content_7a105f548d9c6916,True,False,True,<NA>,125,1,616,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>
3,2026-03-01,client_73cda7b4e4f265ea,content_905aa32a0230694e,True,False,True,<NA>,7,0,28,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>
4,2026-03-01,client_73cda7b4e4f265ea,content_a3ea9792f793ec72,True,False,True,<NA>,11,0,25,...,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>,<NA>


In [38]:
con.sql("""
DESCRIBE
SELECT *
FROM read_parquet('/content/march_2026.parquet')
""").df()

,column_name,column_type,null,key,default,extra
0,report_date,DATE,YES,None,None,None
1,client_hash_id,VARCHAR,YES,None,None,None
2,content_hash_id,VARCHAR,YES,None,None,None
3,client_has_gsc,BOOLEAN,YES,None,None,None
4,client_has_ga4,BOOLEAN,YES,None,None,None
5,gsc_data_available,BOOLEAN,YES,None,None,None
6,ga4_data_available,BOOLEAN,YES,None,None,None
7,gsc_impressions,BIGINT,YES,None,None,None
8,gsc_clicks,BIGINT,YES,None,None,None
9,gsc_sum_position,BIGINT,YES,None,None,None


In [39]:
con.sql("""
DESCRIBE
SELECT *
FROM read_parquet('/content/march_2026.parquet')
""").df()

,column_name,column_type,null,key,default,extra
0,report_date,DATE,YES,None,None,None
1,client_hash_id,VARCHAR,YES,None,None,None
2,content_hash_id,VARCHAR,YES,None,None,None
3,client_has_gsc,BOOLEAN,YES,None,None,None
4,client_has_ga4,BOOLEAN,YES,None,None,None
5,gsc_data_available,BOOLEAN,YES,None,None,None
6,ga4_data_available,BOOLEAN,YES,None,None,None
7,gsc_impressions,BIGINT,YES,None,None,None
8,gsc_clicks,BIGINT,YES,None,None,None
9,gsc_sum_position,BIGINT,YES,None,None,None


In [40]:
df = con.sql("""
SELECT *
FROM read_parquet('/content/march_2026.parquet')
LIMIT 1
""").df()

print(df.columns.tolist())

['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events']


In [41]:
from huggingface_hub import whoami

print(whoami())

{'type': 'user', 'id': '6a4fccc2c410d0c5682392e7', 'name': 'khadijamajeed', 'fullname': 'Khadija Majeed', 'email': 'khadijamajeed838@gmail.com', 'emailVerified': True, 'canPay': False, 'billingMode': 'prepaid', 'periodEnd': 1788220800, 'isPro': False, 'avatarUrl': '/avatars/d0b2a221671f5aa502c04545b379809f.svg', 'orgs': [], 'auth': {'type': 'access_token', 'accessToken': {'displayName': 'HF_TOKEN', 'role': 'read', 'createdAt': '2026-07-28T10:07:51.647Z'}}}


In [42]:
from huggingface_hub import HfApi

api = HfApi()

files = api.list_repo_files(
    repo_id="FlyRank/internship-warehouse",
    repo_type="dataset"
)

print(files)

['.gitattributes', 'README.md', 'dim_clients.parquet', 'dim_content.parquet', 'fact_content_daily_performance/month=2025-01/data_0.parquet', 'fact_content_daily_performance/month=2025-02/data_0.parquet', 'fact_content_daily_performance/month=2025-03/data_0.parquet', 'fact_content_daily_performance/month=2025-04/data_0.parquet', 'fact_content_daily_performance/month=2025-05/data_0.parquet', 'fact_content_daily_performance/month=2025-06/data_0.parquet', 'fact_content_daily_performance/month=2025-07/data_0.parquet', 'fact_content_daily_performance/month=2025-08/data_0.parquet', 'fact_content_daily_performance/month=2025-09/data_0.parquet', 'fact_content_daily_performance/month=2025-10/data_0.parquet', 'fact_content_daily_performance/month=2025-11/data_0.parquet', 'fact_content_daily_performance/month=2025-12/data_0.parquet', 'fact_content_daily_performance/month=2026-01/data_0.parquet', 'fact_content_daily_performance/month=2026-02/data_0.parquet', 'fact_content_daily_performance/month=20

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

One row represents the daily performance of one content item for one client on one report date.

The development window is March 2026, using the `fact_content_daily_performance` March partition.

The observed time window is 2026-03-01 through 2026-03-31, with 9,841,378 rows.

The analysis focuses on content-level search and engagement performance across clients.

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

### Features

The initial features are:

- `gsc_impressions` — Google Search Console search visibility.
- `gsc_clicks` — Google Search Console search clicks.
- `ga4_pageviews` — pageviews measured in GA4.
- `sessions_organic` — organic search sessions.
- `scroll_events` — observed engagement through scroll activity.

These are measurable performance signals rather than identifiers.

### Label / outcome proxy

For the leakage experiment, `gsc_clicks` is treated as an outcome proxy and a label-derived version of it is intentionally introduced as a leaked feature.

The leaked feature is not retained in the final feature set.

### Context

- `report_date`
- `client_hash_id`
- `content_hash_id`
- `client_has_gsc`
- `client_has_ga4`
- `gsc_data_available`
- `ga4_data_available`

These fields describe the observation, client/content identity, or whether a measurement source is available.

### Excluded

The identifier fields `client_hash_id` and `content_hash_id` are excluded from the numeric feature set because they identify entities but do not represent measurable performance signals.

The raw date field `report_date` is also excluded from the numeric feature set because it defines the observation window rather than being a performance measurement.

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

In [43]:
con.sql("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT report_date || '|' || client_hash_id || '|' || content_hash_id) AS unique_grain_keys
FROM read_parquet('/content/march_2026.parquet')
""").df()

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

,total_rows,unique_grain_keys
0,9841378,9841378


In [44]:
con.sql("""
SELECT
    COUNT(*) AS row_count,
    MIN(report_date) AS start_date,
    MAX(report_date) AS end_date
FROM read_parquet('/content/march_2026.parquet')
""").df()

,row_count,start_date,end_date
0,9841378,2026-03-01,2026-03-31


In [45]:
con.sql("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(*) FILTER (WHERE gsc_data_available IS TRUE) AS gsc_available_rows,
    COUNT(*) FILTER (WHERE ga4_data_available IS TRUE) AS ga4_available_rows
FROM read_parquet('/content/march_2026.parquet')
""").df()

,total_rows,gsc_available_rows,ga4_available_rows
0,9841378,3611061,413966


In [46]:
df = con.sql("""
SELECT *
FROM read_parquet('/content/march_2026.parquet')
LIMIT 1
""").df()

print("\n".join(df.columns.tolist()))

report_date
client_hash_id
content_hash_id
client_has_gsc
client_has_ga4
gsc_data_available
ga4_data_available
gsc_impressions
gsc_clicks
gsc_sum_position
gsc_avg_position
ga4_pageviews
ga4_sessions
ga4_users
ga4_engaged_sessions
ga4_total_engagement_sec
sessions_organic
sessions_direct
sessions_referral
sessions_social
sessions_paid
sessions_ai
ai_chatgpt
ai_perplexity
ai_gemini
ai_copilot
ai_claude
ai_meta
ai_other
scroll_events


In [47]:
features = con.sql("""
SELECT
    gsc_impressions,
    gsc_sum_position,
    ga4_pageviews,
    sessions_organic,
    scroll_events,
    gsc_clicks AS label
FROM read_parquet('/content/march_2026.parquet')
WHERE gsc_data_available IS TRUE
""").df()

features.head()

,gsc_impressions,gsc_sum_position,ga4_pageviews,sessions_organic,scroll_events,label
0,20,67,<NA>,<NA>,<NA>,0
1,1,0,<NA>,<NA>,<NA>,0
2,125,616,<NA>,<NA>,<NA>,1
3,7,28,<NA>,<NA>,<NA>,0
4,11,25,<NA>,<NA>,<NA>,0


### Why these features are available at the decision moment

- `gsc_impressions`: available when GSC data is available for the observation.
- `gsc_sum_position`: available when GSC data is available for the observation.
- `ga4_pageviews`: available when GA4 data is available for the observation.
- `sessions_organic`: available when the relevant analytics data is available.
- `scroll_events`: available when engagement tracking is available.

These are treated as observed measurements rather than guaranteed measurements for every row because data-source availability is uneven.

In [48]:
leakage = features.copy()

leakage["leaked_clicks"] = leakage["label"]

print("Rows:", len(leakage))
print("Leaked feature created:", "leaked_clicks" in leakage.columns)
print(
    "Mean absolute error using leaked_clicks:",
    (leakage["label"] - leakage["leaked_clicks"]).abs().mean()
)
honest_features = leakage.drop(columns=["leaked_clicks"])

print(honest_features.columns.tolist())

Rows: 3611061
Leaked feature created: True
Mean absolute error using leaked_clicks: 0.0
['gsc_impressions', 'gsc_sum_position', 'ga4_pageviews', 'sessions_organic', 'scroll_events', 'label']


### Leakage experiment

I intentionally created `leaked_clicks` by copying the outcome proxy `gsc_clicks` into a feature column. This produced a mean absolute error of 0 because the feature contained the outcome itself.

This is leakage because the model would effectively be given the answer. The leaked feature is therefore removed from the final feature frame.

The final feature set does not contain `leaked_clicks`.

## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

The March 2026 data does not have complete measurement coverage across all rows.

There are 9,841,378 total rows. GSC data is available for 3,611,061 rows, while GA4 data is available for 413,966 rows. Therefore, analyses using these sources are based on the available subset and should not be treated as complete coverage.

The data shows observed associations and performance measurements. It cannot by itself establish that one feature caused a change in performance, so results should be treated as directional decision-support rather than causal evidence.

The March 2026 slice is used for development. It should not be treated as evidence of performance on the later sealed/test period.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.